test/test_util_network.cpp

```diff
@@ -152,6 +153,46 @@ TEST(Network_PostOperation)
 }
 
 
+TEST(Network_RunUntilStopped)
+{
+    network::Service service;
+    auto post_to_service = [&] {
+        auto [promise, future] = util::make_promise_future<void>();
+        service.post([promise = std::move(promise)](Status s) mutable {
+            if (!s.is_ok()) {
+                promise.set_error(s);
+                return;
+            }
+            promise.emplace_value();
+        });
+        return std::move(future);
+    };
+    auto before_run = post_to_service();
+
+    auto [thread_stopped_promise, thread_stopped_future] = util::make_promise_future<void>();
+    std::thread thread([&service, promise = std::move(thread_stopped_promise)]() mutable {
+        service.run_until_stopped();
+        promise.emplace_value();
+    });
+
+    before_run.get();
+    CHECK_NOT(thread_stopped_future.is_ready());
+    // Post while it's running. This should get fulfilled immediately.
+    post_to_service().get();
```

Comment:
Can you add an extra step to sleep here for 100 ms (for the events to clear out in the run loop) and then create a timer that waits for a short time (e.g. 250 ms) to check to verify these run as well?
I haven't traced the whole flow of events in the Service, but it looks like new timers get added to the `m_wait_operations` queue, but the `kevent()` request in `wait_and_activate()` is set to timeout after 24 hours if there are no timers set (i.e. timeout is 0). I haven't been able to find where/if the kevent() queue is interrupted when a new timer is added.

Reply:
So having thought about this a bit more, I think the only safe thing to do outside the event loop is call `post()`. I think in practice that means all timers get added from inside the event loop, and so when whatever function that registers the timer returns, the call to `run()` hasn't returned, and therefore there's no need to interrupt the loop because you're already inside it.